File: 79WordSearch.java
class Solution {
    public boolean exist(char[][] board, String word) {
        Map<String, Integer> map = new HashMap<>();
        Queue<int[]> q = new LinkedList<>();
        int wordDepthMax = 0;

        for (int x=0; x < board.length; x++) {
            for (int y=0; y < board[0].length; y++) {
                int[] pos = new int[]{x, y};
                if (word.charAt(0) == board[x][y]
                && !map.containsKey(Arrays.toString(pos))) {
                    q.add(pos);
                    map.put(Arrays.toString(pos), 0);
                    if (word.length() == bfs(pos, board, word, map, q)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private int bfs(int[] pos, char[][] board, String word, Map<String, Integer> map, Queue<int[]> q) {
        int wordDepthMax = 0;
        while (!q.isEmpty()) {
            int[] elem = q.remove();
            int x = elem[0], y = elem[1];
            int left = x > 0 ? x - 1 : x;
            int right = x < board.length - 1 ? x + 1 : x;
            int up = y > 0 ?  y - 1 : y;
            int down = y < board[0].length - 1 ? y + 1 : y;
            for (int dirx = left; dirx <= right; dirx++) {
                for (int diry = up; diry <= down; diry++) {
                    int wordPos = map.get(Arrays.toString(elem))+1;
                    wordDepthMax = Math.max(wordDepthMax, wordPos);
                    int[] newPos = new int[]{dirx, diry};

                    if (word.charAt(wordPos) == board[dirx][diry]
                    && !map.containsKey(Arrays.toString(newPos))) {
                        q.add(newPos);
                        map.put(Arrays.toString(newPos), wordPos);
                    }
                }
            }
        }
        return wordDepthMax;
    }
}

File: 79WordSearch.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        unordered_set<char> set;
        vector<vector<int>> visited(board.size(), vector<int>(board[0].size(), 0));
        
        for (char c : word) {
            set.insert(c);
        }
        
        for (int i=0; i<board.size(); i++) {
            for (int u=0; u<board[i].size(); u++) {
                if (parseGrid(board, word, i, u, 0, visited, set)) {
                    return true;
                }
            }
        }
        
        return false;
    }
    
    bool parseGrid(vector<vector<char>>& board, string word, int i, int u, int wordIdx, vector<vector<int>>& visited, unordered_set<char>& set) {
        if (wordIdx == word.length())
            return true;
        
        if (i<0 || i>=board.size() || u<0 || u>=board[0].size() || word[wordIdx] != board[i][u] || visited[i][u]) {
            return false;
        }
        
        visited[i][u] = true;
        
        if (parseGrid(board, word, i-1, u, wordIdx + 1, visited, set)
           || parseGrid(board, word, i+1, u, wordIdx + 1, visited, set)
           || parseGrid(board, word, i, u-1, wordIdx + 1, visited, set)
           || parseGrid(board, word, i, u+1, wordIdx + 1, visited, set)) {
            return true;
        }
        
        if (set.find(board[i][u]) != set.end())
            visited[i][u] = false;
        
        return false;
    }
};